src/EditorFeatures/Core/Implementation/AutomaticCompletion/AbstractEditorBraceCompletionSessionFactory.cs

```diff
@@ -2,48 +2,34 @@
+using System.Collections.Immutable;
+using System.Linq;
 using System.Threading;
+using Microsoft.CodeAnalysis.BraceCompletion;
+using Microsoft.CodeAnalysis.Editor.Implementation.AutomaticCompletion.Sessions;
 using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
-using Microsoft.CodeAnalysis.LanguageServices;
-using Microsoft.CodeAnalysis.Shared.Extensions;
 
 namespace Microsoft.CodeAnalysis.Editor.Implementation.AutomaticCompletion
 {
     internal abstract class AbstractEditorBraceCompletionSessionFactory : ForegroundThreadAffinitizedObject, IEditorBraceCompletionSessionFactory
     {
-        protected AbstractEditorBraceCompletionSessionFactory(IThreadingContext threadingContext)
-            : base(threadingContext)
-        {
-        }
-
-        protected abstract bool IsSupportedOpeningBrace(char openingBrace);
-
-        protected abstract IEditorBraceCompletionSession CreateEditorSession(Document document, int openingPosition, char openingBrace, CancellationToken cancellationToken);
+        private readonly ImmutableArray<IEditorBraceCompletionSession> _braceCompletionSessions;
 
-        public IEditorBraceCompletionSession TryCreateSession(Document document, int openingPosition, char openingBrace, CancellationToken cancellationToken)
+        protected AbstractEditorBraceCompletionSessionFactory(
+            IEnumerable<IBraceCompletionService> braceCompletionServices,
+            IThreadingContext threadingContext)
+            : base(threadingContext)
         {
-            this.AssertIsForeground();
-
-            if (IsSupportedOpeningBrace(openingBrace) &&
-                CheckCodeContext(document, openingPosition, openingBrace, cancellationToken))
-            {
-                return CreateEditorSession(document, openingPosition, openingBrace, cancellationToken);
-            }
-
-            return null;
+            _braceCompletionSessions = braceCompletionServices
+                .Select(service => (IEditorBraceCompletionSession)new EditorBraceCompletionSession(service))
+                .ToImmutableArray();
         }
 
-        protected virtual bool CheckCodeContext(Document document, int position, char openingBrace, CancellationToken cancellationToken)
+        public IEditorBraceCompletionSession? TryCreateSession(Document document, int openingPosition, char openingBrace, CancellationToken cancellationToken)
         {
             this.AssertIsForeground();
-
-            // check that the user is not typing in a string literal or comment
-            var tree = document.GetSyntaxRootSynchronously(cancellationToken).SyntaxTree;
-            var syntaxFactsService = document.GetLanguageService<ISyntaxFactsService>();
-
-            return !syntaxFactsService.IsInNonUserCode(tree, position, cancellationToken);
+            return _braceCompletionSessions.SingleOrDefault(session => session.IsValidForBraceCompletion(openingBrace, openingPosition, document, cancellationToken));
```

Comment:
SingleOrDefault is dangerous in hte case where two sessions say they're valid.  we'll throw in that case.  is that appropriate here?

Reply:
yes - there should only ever be one session, if there isn't it's a bug that we should know about.  I could nfw if preferred rather than throwing